### ISSUE TITLE
windows+fltk1.3.x+1.4.x: "fluid -c somefile.fl" fails without error if .fl file unreadable

### ISSUE THREAD

**Issue by erco77** [fltk/fltk#224]
Toolchain: Visual Studio 2017 + fltk 1.3.5 (built as "Release", not "Debug")

If one uses 'fluid -c' in a Makefile on Windows to build a project's .fl file -> .cpp/.h, if fluid encounters a problem reading the .fl file (e.g. due to permissions), it just fails silently with a non-zero exit code. When used in the context of a Makefile, it's unclear what the problem is.

```
C:\somepath>type noperm.fl
Access is denied.

C:\somepath>fluid -c noperm.fl
                              <-- no error message
C:\somepath>echo %ERRORLEVEL%
1
```
If just `fluid noperm.fl` is executed, it properly shows an error dialog: "Can't read noperm.fl: Permission denied".

**Comment by erco77** [comment]
I see the same problem if the output files have no permission.

I think the issue here is in "Release" mode, fluid is running as a "Windows" application, making it unreliable for use in a Makefile because it (a) backgrounds itself, and (b) disconnects from the DOS console.

**Comment by Albrecht-S** [comment]
> The workaround is to change this to /SUBSYSTEM:CONSOLE ...

Greg, can you please test the attached patch [fluid_console.diff.txt](https://github.com/fltk/fltk/files/6384291/fluid_console.diff.txt) against FLTK 1.4. Removing the "WIN32" flag builds fluid with `/SUBSYSTEM:CONSOLE`. The same patch does not apply cleanly to 1.3 but it's easy to adapt if you need (the change is in line 66 vs. 69, but context is different).

I tested this patch successfully in a "git bash" console and a standard `DOS window`. `fluid` does no longer background itself and you can see the error message.

**The test I'm asking you to do is meant as a proof of concept, not a final solution**. I can only simulate what you are doing, I don't use Makefiles to build on Windows (except with MinGW/MSYS which doesn't seem to be concerned).

> I'm wondering how fluid generates FLTK's own .cxx/h files reliably during library builds without racing the compiler for the generation of the .cxx/.h files. Also, I imagine getting proper exit codes from fluid is affected by this.

Interesting question! I see what you mean (backgrounding, no exit code) but I don't know a good answer. I've never seen issues though. I could imagine that you'd need to run the build twice if it failed because the fluid file gets written "too late" for the build process (race) but I've never seen such issues in my builds.


It seems clear that it's not feasible to build `fluid` always as a console app because it would create the console window when run from the file explorer. In the "_worst case_" we could build two different executables (`fluid` and `fluidc`?). Making the decision a build option would be possible but likely not what we want (users would need two **complete builds** to get fluid for commandline and for interactive use).

**Comment by Albrecht-S** [comment]
Additional question if we wanted to build two fluid executables: which (Windows) build systems would be affected? Would it only be when building with Visual Studio (MSVC) or are other Windows builds also affected by this "console" issue? Maybe all Windows builds?

**Comment by erco77** [comment]
OK, I'll try it later today if this second Pfizer shot doesn't knock me out.

In this all-windows app I've been working around which has the same issue; needs to provide both console and GUI interaction, I made a separate executable; one for windows called myapp, the other for console use called myapp-cmd.

It worked well because that way for sure the backend modules/classes had NO gui code mixed in; all gui code was in the myapp.cpp, and the console program myapp-cmd.cpp was pure stdin/out/err, not one line of gui code, and could compile without any of the FLTK libraries and their OS dependencies. Any interaction between the low level classes and GUI were done with messaging using a simple class.
This ensures the backend modules truly are modular and free of any GUI logic madness. I think fluid suffers from this a bit; there's a lot of GUI code mixed into the backend.

**Comment by erco77** [comment]
OK, found a minute to apply your patch to 1.4.x git. Short answer: It works!
Shows the error in the console, and definitely does not background itself. (Easy to test by giving it a UNC path to a nonexistent file server, e.g. `fluid -c \\google.org\some\where.fl` which causes fluid to hang trying to find it)

FWIW, here's what I did after applying the patch:

First I created a new build dir, cd'ed to it, and ran:
```
C:\\fltk-1.4.x-git\build-albrecht-vs2017>cmake -D CMAKE_BUILD_TYPE=Release ..
```
Side note: not sure what this warning was about, but I think I always see it. I did run cmake from the "VS 2017 DOS Command Prompt" window:
```
[..]
-- Check size of long long - done
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
-- Warning: Header file GL/glu.h was not found.         <<<<
-- The FLTK team recommends to rerun CMake from a       <<<<
-- "Developer Command Prompt for Visual Studio xxxx"    <<<<
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
-- Found OpenGL: opengl32
[..]
```
Then loaded fltk.sln into VS 2017's IDE (which takes about 5mins.. I guess 8gig of ram is apparently not enough for Win8+VS)

Side note: even though CMAKE_BUILD_TYPE was set to "Release" on the cmake command line, still had to change the VS IDE pulldown menu from Debug -> Release (which on my slow machine takes around 5 minutes to complete, lol)

Trying not to get sidetracked, built the entire solution, then tested the generated fluid on a file I didn't have permission to read, and it gave a proper error message.

**Comment by erco77** [comment]
I should add that just changing the Runtime type to /SUBSYSTEM:CONSOLE for fluid worked as well, which is what I did last night.

**Comment by Albrecht-S** [comment]
> Hmm, does that patch change the /SUBSYSTEM in the IDE to 'CONSOLE'?
I ask because last night I just changed /SUBSYSTEM to console to get things working.

Last things first: I wrote in my [previous message with the patch](https://github.com/fltk/fltk/issues/224#issuecomment-827573895):
> Removing the "WIN32" flag builds fluid with /SUBSYSTEM:CONSOLE.

You wrote:
> just surprised the WIN32 flag affected that.

Yes, that's its sole purpose in CMake (add_executable). The naming is questionable, but that's what it does. It sets all build flags necessary to build a GUI (`/SUBSYSTEM:WINDOWS`) program on Windows. Leaving it away makes it `/SUBSYSTEM:CONSOLE`. There's a similar flag for macOS bundles, btw.

**Comment by erco77** [comment]
Regarding your follow up questions:

> which (Windows) build systems would be affected?

I'm pretty sure ANY windows builds are affected, as this is an OS level thing. Which is to say even mingw, which has a slightly differently named flag (-console?), but has the same effect IIRC.

> we could build two different executables (fluid and fluidc?).

Yes, a separate exec seems the only solution I think. See my reply above regarding details. I think fluid-cmd would be clearer; this is what I've seen other commercial apps do that run into the same problem, and I've used it too.
The only way to avoid a name change is to make separate directories, e.g. windows/fluid.exe and console/fluid.exe, but that can lead to confusion I think.

Note there are ways in DOS to start a windows app and wait for it to finish and get its exit code, namely:
```
    start /WAIT fluid -c bla.fl
```
(But that doesn't solve the stdin/out/error problem.)

Re: your follow up comment to me about WIN32 affecting SUBSYSTEM, ya my edit and your reply crossed in time; I read back your first response and saw that was the effect.

**Comment by Albrecht-S** [comment]
Regarding the CMake warnings:
```
-- Warning: Header file GL/glu.h was not found.         <<<<
-- The FLTK team recommends to rerun CMake from a       <<<<
-- "Developer Command Prompt for Visual Studio xxxx"    <<<<
```
This is only a recommendation, but I think it's not necessary. The reason is that `GL/glu.h` may not be found by `CMake` but is nevertheless available somewhere in the Windows SDK. This ought to be found when you run CMake from a developer command prompt but meanwhile I think we should remove it because the build works fine anyway. Magic.

> Side note: even though CMAKE_BUILD_TYPE was set to "Release" on the cmake command line, still had to change the VS IDE pulldown menu from Debug -> Release ...

FYI: `CMAKE_BUILD_TYPE` is **ignored** in VS build generators because these are "multi-config" generators, i.e. you can select the build type in the IDE. It's only relevant to set a particular build type for Makefiles, Ninja, and other single-config generators. I'd like it to set the **default build type** for MSVC rather than being ignored but that's IIRC something the CMake guys can't do (or don't want to do) for some reason.

**Comment by Albrecht-S** [comment]
> I'm pretty sure ANY windows builds are affected, as this is an OS level thing. Which is to say even mingw, which has a slightly differently named flag (-console?), but has the same effect IIRC.

This would be `-mwindows` and `-mconsole` IIRC. I noticed a different behavior when I ran the GUI version of fluid in a MinGW shell though: stdout seemed to work which surprised me.

But I agree that it would make sense to build two executables for **all** Windows builds.

> I think fluid-cmd would be clearer; this is what I've seen other commercial apps do that run into the same problem, and I've used it too.

No problem, fine for me. Side note: the FLTK docs describe this backgrounding of fluid somewhere, checked yesterday, but I don't have a link.

> start /WAIT fluid -c bla.fl

Oh yes, there's something in my faint Windows memory... That's probably the way CMake runs these commands, it makes sense.

**Comment by Albrecht-S** [comment]
FTR: I added label `CMake` because we need _CMake_ additions to build two executables for Windows, but not **only** CMake. We'll need to:

- add CMake support for two fluid executables
- add configure/make support as well
- change `fluid` commands to use the commandline executable

Everything for Windows only, of course.

**Comment by Albrecht-S** [comment]
I'm now working on this and so far I have a proof of concept (needs some more tests).

1. Generate two executables on all Windows builds (`fluid.exe` and `fluid-cmd.exe`) as proposed.
2. Do 1. also if cross-compiling.
3. Use `fluid-cmd.exe` to generate `.cxx` and `.h` files from `.fl` files.
4. Install both `fluid` executables if [make] `install` is requested.
5. For now I only changed the `CMake` build system.
6. I'm not sure yet if I can do this for `configure/make` as well (let's see later ...).

This should hopefully resolve both this issue and issue #293.

Regarding the DOS window popup: in my tests I couldn't see windows popping up but others may need to test this (on slower systems?).

**Comment by Albrecht-S** [comment]
I'm now working on this and so far I have a proof of concept (needs some more tests).

1. Generate two executables on all Windows builds (`fluid.exe` and `fluid-cmd.exe`) as proposed.
2. Do 1. also if cross-compiling.
3. Use `fluid-cmd.exe` to generate `.cxx` and `.h` files from `.fl` files.
4. Install both `fluid` executables if [make] `install` is requested.
5. For now I only changed the `CMake` build system.
6. I'm not sure yet if I can do this for `configure/make` as well (let's see later ...).

This should hopefully resolve both this issue and issue #293.

Regarding the DOS window popup: in my tests I couldn't see windows popping up but others may need to test this (on slower systems?).

**Comment by Albrecht-S** [comment]
@erco77 Greg, can you please test the patch, either by cloning/pulling [from my fork: branch `fluid-cmd`](https://github.com/Albrecht-S/fltk/tree/fluid-cmd) or use the attached patch file. I tried to test as much as possible but you may be able to test more specific for your application and/or build environment.

[issue-224.patch.txt](https://github.com/fltk/fltk/files/8403351/issue-224.patch.txt)

Note: please use `CMake` to build FLTK (`configure/make` support is not yet included).

**Comment by erco77** [comment]
OK, will try to follow up today.

**Comment by erco77** [comment]
Seems to work well for my build case; I created existing .cxx and .h files with tight permissions so that fluid-cmd wouldn't be able to overwrite them (which was what caused me to submit this issue), and it's showing a useful error now in the DOS build window and the exit code from fluid-cmd is non-zero on error, flagging to nmake that it failed, which is all good.
Then when I fixed the perms on those files and re-built, it worked as expected; no errors from fluid-cmd, and a zero exit code.

```
Z:\erco\company\XXX\XXXFirmwareUpdater> Z:\tmp\fltk-albrecht-fork\build-vs2019-nmake\bin\fluid-cmd.exe -c Application_UI.fl
Application_UI.cpp and Application_UI.h : Permission denied

Z:\erco\company\XXX\XXXFirmwareUpdater>echo %ERRORLEVEL%
1
```

**Comment by Albrecht-S** [comment]
@erco77 Thanks for testing and feedback.

I declare this issue fixed (commit 228d04d0e0d6e968c59bcab606e609eb1103f917).

I'm not going to try to build `fluid-cmd` for Windows using `configure/make`. If this is needed then someone may open another issue and/or provide a patch.
